### PR TITLE
[CARBONDATA-4080] Wrong results for select count on invalid segments

### PR DESCRIPTION
**Pull request by akkio-97**
### Why is this PR needed?

Wrong results for 
- select count on marked for delete segment 
- select count on compacted segment 

Issue comes only when the user explicitly sets deleted/compacted segments using the property carbon.input.segments.
As select * on such segments gives 0 rows as output, in order to maintain consistency, select count should also give 0 rows.
 
 ### What changes were proposed in this PR?

So this PR focuses on considering only valid segments to be accessed during select count query.

    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?

 - Yes

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5124/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3362/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3367/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5129/

**Comment by akkio-97** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5130/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3368/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5132/

**Comment by akkio-97** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5133/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3371/

**Comment by kunal642** [comment]
@akashrn5 Please check whether this fix will be ok for SI and MV.

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala [hunk 1]
```diff
@@ -420,4 +421,40 @@ class TestSegmentReading extends QueryTest with BeforeAndAfterAll {
...
+      "create table carbonTable(a string, b int, c string) stored as carbondata tblproperties" +
+      "('carbon.clean.file.force.allowed' = true)")
+    sql("insert into carbonTable values ('k',1,'k'), ('k',1,'b')")
+    sql("insert into carbonTable values ('a',2,'a')")
+    sql("insert into carbonTable values ('b',2,'b'),('b',2,'b')")
+    sql("insert into carbonTable values ('c',2,'c')")
+
+    sql("delete from table carbonTable where segment.id in (0,3)")
+    sql("set carbon.input.segments.default.carbonTable = 0,2,3")
+    sql("show segments for table carbonTable").show()
```
Comment (kunal642) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala:437]: please remove this

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala [hunk 2]
```diff
@@ -420,4 +421,40 @@ class TestSegmentReading extends QueryTest with BeforeAndAfterAll {
...
+      "create table carbonTable(a string, b int, c string) stored as carbondata tblproperties" +
+      "('carbon.clean.file.force.allowed' = true)")
+    sql("insert into carbonTable values ('k',1,'k'), ('k',1,'b')")
+    sql("insert into carbonTable values ('a',2,'a')")
+    sql("insert into carbonTable values ('b',2,'b'),('b',2,'b')")
+    sql("insert into carbonTable values ('c',2,'c')")
+    sql("alter table carbonTable compact 'major'")
+
+    sql("set carbon.input.segments.default.carbonTable = 0,1,2,3,0.1")
+    sql("show segments for table carbonTable").show()
```
Comment (kunal642) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala:455]: please remove this

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala (hunk 1 above)
Comment (akkio-97) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala]: done

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala (hunk 2 above)
Comment (akkio-97) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala]: done

**Comment by vikramahuja1001** [comment]
@akkio-97 , please add a test case for partition table as well

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala [hunk 3]
```diff
@@ -420,4 +421,69 @@ class TestSegmentReading extends QueryTest with BeforeAndAfterAll {
...
+    sql("insert into carbonTable values ('a',1,'a')")
+    sql("insert into carbonTable values ('b',2,'b'),('b',2,'b')")
+    sql("insert into carbonTable values ('c',3,'c')")
+
+    sql("drop index if exists indextable1 on carbonTable")
+    sql("create index indextable1 on table carbonTable (c) AS 'carbondata'")
+
+    sql("alter table carbonTable compact 'major'")
+    sql("set carbon.input.segments.default.carbonTable = 0,1,2,3,0.1")
+    checkAnswer(sql("select count(*) from carbonTable"), Seq(Row(6)))
```
Comment (Indhumathi27) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala:486]: add a query check which has filter on SI column

**Comment by akashrn5** [comment]
> @akashrn5 Please check whether this fix will be ok for SI and MV.

for SI , better to have a test case and for MV during query if set segments is done, then it will not hit query, but please add a test case which set segments and create MV. Please check in MVCreatetest class and add

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5136/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3374/

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala (hunk 3 above)
Comment (akkio-97) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/segmentreading/TestSegmentReading.scala]: done

**Comment by akkio-97** [comment]
> @akkio-97 , please add a test case for partition table as well

done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5140/

**Comment by akkio-97** [comment]
> > @akashrn5 Please check whether this fix will be ok for SI and MV.
> 
> for SI , better to have a test case and for MV during query if set segments is done, then it will not hit query, but please add a test case which set segments and create MV. Please check in MVCreatetest class and add

done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5142/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3380/

**Comment by kunal642** [comment]
LGTM

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5197/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3438/

**Pull request closed without merging**